Bluetooth: Mesh: disable/enable advertising when suspending/resuming Mesh

HaavardRei opened the pull request:
- Adds `bt_mesh_adv_disable()` to the advertising API for Mesh.
- Disables/enables advertising when suspending/resuming Mesh.
- Adds a bsim test checking that periodic publication is stopped when Mesh is suspended.

HaavardRei commented:
Adding the implemented `bt_mesh_adv_disable()` to `bt_mesh_suspend()` broke the `pb_remote_timeout.sh` test in `test_provision.c` [(here)](https://github.com/zephyrproject-rtos/zephyr/blob/1e15949fe04a6f778f2e0f32f62ef51fc22ef66d/tests/bsim/bluetooth/mesh/src/test_provision.c#L159). As far as I can see the changes should not be in conflict with the test should work, but likely disabling the advertising instance(s) causes some problems in the test. I will await some reviews before investigating further.
DNM was added and will be removed when this is resolved.

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+
+conf=prj_mesh1d1_conf
+overlay="overlay_gatt_conf_overlay_psa_conf"
+RunTest mesh_suspend_test_psa \
```
```suggestion
RunTest mesh_suspend_resume_disable_psa \
```

akredalen commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+
+# Test that periodic publication is stopped when suspending Mesh and disabling
+# Bluetooth, and that it is started again when Bluetooth is re-enabled and Mesh is resumed.
+# The test will fail if no publication is received, or if a publication is received when Mesh
```
Suggestion to provide clarity:  This test will fail under two conditions: if no publications are received while the Mesh is enabled, or if a publication is received while Mesh is disabled.

HaavardRei replied:
Clarified! Thanks for input!

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_resume.sh:
```diff
@@ -0,0 +1,29 @@
+# 3. Resume Mesh 20 seconds after suspension.
+
+overlay=overlay_gatt_conf
+RunTest mesh_suspend_test \
```
```suggestion
RunTest mesh_suspend_resume \
```

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_resume.sh:
```diff
@@ -0,0 +1,29 @@
+
+conf=prj_mesh1d1_conf
+overlay=overlay_gatt_conf
+RunTest mesh_suspend_test \
```
```suggestion
RunTest mesh_suspend_resume_1d1 \
```

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_resume.sh:
```diff
@@ -0,0 +1,29 @@
+
+conf=prj_mesh1d1_conf
+overlay="overlay_gatt_conf_overlay_psa_conf"
+RunTest mesh_suspend_test_psa \
```
```suggestion
RunTest mesh_suspend_resume_psa \
```

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_resume.sh:
```diff
@@ -0,0 +1,29 @@
+
+# Test that periodic publication is stopped when suspending Mesh, and that it
+# is started again when Mesh is resumed. The test will fail if no publication is
+# received, or if a publication is received when Mesh is disabled.
```
```suggestion
# received, or if a publication is received when Mesh is suspended.
```

alxelax commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+# Test that periodic publication is stopped when suspending Mesh and disabling
+# Bluetooth, and that it is started again when Bluetooth is re-enabled and Mesh is resumed.
+# The test will fail if no publication is received, or if a publication is received when Mesh
+# is disabled.
```
```suggestion
# is suspended.
```

akredalen commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+# The test will fail if no publication is received, or if a publication is received when Mesh
+# is disabled.
+#
+# Test procedure:
```
This test procedure might be missing what is checked during and after suspending and resuming mesh. Same for the other test script

HaavardRei replied:
Added some more info.

PavelVPV commented on tests/bsim/bluetooth/mesh/src/main.c:
```diff
@@ -19,6 +19,7 @@ extern struct bst_test_list *test_sar_pst_install(struct bst_test_list *test);
 #endif /* defined(CONFIG_BT_MESH_V1d1) */
 #elif defined(CONFIG_BT_MESH_GATT_PROXY)
 extern struct bst_test_list *test_adv_install(struct bst_test_list *test);
+extern struct bst_test_list *test_suspend_install(struct bst_test_list *test);
```
I agree, GATT advertising suspension should be added as well and it should be checked with at least one device connected over GATT to check if proxy connections are cleared after `bt_disable()` call.

omkar3141 replied:
May be it is worth to do this as a separate task? Just to not explode the scope if we discover that proxy related modules require rework for clean suspend-resume transitions.

HaavardRei replied:
I moved to use the "basic" configs, then we can do this as a separate task.

omkar3141 commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -225,6 +225,12 @@ int bt_mesh_adv_enable(void)
 	return 0;
 }
 
+int bt_mesh_adv_disable(void)
```
I'm little concerned about the naming here. The function says "_adv_disable()", and internally calls "_thread_suspend()". These two are slightly different operations when you look at these words - disable vs suspend. May be to avoid confusions in future, it may be nicer to have API named as "bt_mesh_adv_suspend()".

HaavardRei replied:
I agree, but there is no `thread_disable()` API, and adding a `_adv_disable()` function requires adding the function both to the legacy and extended advertiser. A possible solution could be to rename both to suspend, but this might pose some issues with how we want the functionality to be with the extended advertising...

omkar3141 commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+LOG_MODULE_REGISTER(LOG_MODULE_NAME, LOG_LEVEL_INF);
+
+#define WAIT_TIME        75 /* seconds */
+#define SUSPEND_START    15 /* seconds */
```
I don't think we do this anywhere else in the stack. Shorter integers are automatically promoted to next higher type, and there is no type shorter than int8/uint8 (except for bool)

HaavardRei replied:
No longer comparing these, see above discussion

omkar3141 commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	/* Ensure that no message is received while Mesh is suspended or disabled. */
+	if (uptime > SUSPEND_START  && uptime < (SUSPEND_START + SUSPEND_DURATION)) {
+		FAIL("Received publication while Mesh is suspended.");
+	}
```
You could add additional `else` clauses here to ensure that at least one publication happens outside of this time block, before suspension and after resumption. If because of some other errors, in suspend-resume mechanism, publish never happens, "test_msg_handler()" will never get called and can potentially create false PASS result.

HaavardRei replied:
With the new use of semaphores, the test will fail if no message is sent/received.

omkar3141 commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	/* Wait until node has published before suspending. */
+	k_sleep(K_SECONDS(SUSPEND_START));
+	err = bt_mesh_suspend();
+	if (err) {
```
I will prefer not. We don't do this in many other tests. It increases vertical span of the code and sometimes makes it difficult to comprehend.

omkar3141 commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+# 3. Enable Bluetooth and resume Mesh 20 seconds after suspension.
+
+overlay=overlay_gatt_conf
+RunTest mesh_suspend_test \
```
I'm bit confused by the word "disable" in the test name. Why are you suggesting to add "disable" there? Isn't this all about suspend-resume only? It seems like we have third API operation of "disabling" mesh stack.

HaavardRei commented:
@omkar3141 I have implemented two tests; the first where we suspend then resume mesh, and the second where we suspend mesh, then **disable** Bluetooth before re-enabling Bluetooth and resuming Mesh.

HaavardRei commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -458,6 +458,13 @@ int bt_mesh_suspend(void)
 
 	bt_mesh_beacon_disable();
 
+	err = bt_mesh_adv_disable();
```
Moved!

HaavardRei commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -225,6 +225,12 @@ int bt_mesh_adv_enable(void)
 	return 0;
 }
 
+int bt_mesh_adv_disable(void)
```
Done, made minor change to resume adv for low latency!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+#include "mesh_test.h"
+
+#include <stddef.h>
+#include <string.h>
```
Removed!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+#define WAIT_TIME        75 /* seconds */
+#define SUSPEND_START    15 /* seconds */
+#define SUSPEND_DURATION 20 /* seconds */
+#define NUM_PUB			  4 /* Times the transmitter will publish. */
```
Fixed!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+{
+	static uint8_t prev_num;
+	uint8_t num = net_buf_simple_pull_u8(buf);
+	uint8_t uptime = k_uptime_get_32() / 1000;
```
Removed use of uptime by replacing with a semaphore-based method

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	/* Wait until node has published before suspending. */
+	k_sleep(K_SECONDS(SUSPEND_START));
```
See above comment

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	.cred_flag = false,
+	.app_idx = 0,
+	.ttl = 5,
+	.period = BT_MESH_PUB_PERIOD_10SEC(1),
```
Done!

HaavardRei commented on tests/bsim/bluetooth/mesh/tests_scripts/suspend/suspend_disable_resume.sh:
```diff
@@ -0,0 +1,30 @@
+# 3. Enable Bluetooth and resume Mesh 20 seconds after suspension.
+
+overlay=overlay_gatt_conf
+RunTest mesh_suspend_test \
```
Fixed, thanks for suggestions!

HaavardRei commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -506,6 +506,26 @@ int bt_mesh_adv_enable(void)
+
+	for (int i = 0; i < ARRAY_SIZE(advs); i++) {
+		err = bt_le_ext_adv_stop(advs[i].instance);
+		if (err) {
```
Done!

HaavardRei commented on subsys/bluetooth/mesh/main.c:
```diff
@@ -488,6 +495,13 @@ int bt_mesh_resume(void)
+	err = bt_mesh_adv_enable();
+	if (err) {
+		LOG_WRN("Re-enabling advertisers failed (err %d)", err);
+		atomic_set_bit(bt_mesh.flags, BT_MESH_SUSPENDED);
```
Reordered!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+#include <zephyr/kernel.h>
+
+#include <zephyr/logging/log.h>
+#define LOG_MODULE_NAME test_suspend
```
Passed directly!

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+{
+	static uint8_t prev_num;
+	uint8_t num = net_buf_simple_pull_u8(buf);
+	uint8_t uptime = k_uptime_get_32() / 1000;
```
Removed use of uptime, see above discussion.

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	}
+	LOG_INF("Mesh resumed.");
+
+	k_sleep(K_SECONDS(35));
```
Added semaphore-based sync instead to receive specified amount of messages.

HaavardRei commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,325 @@
+	}
+
+	/* Wait until node has published before suspending. */
+	k_sleep(K_SECONDS(SUSPEND_START));
```
Removed. (See above discussion)

PavelVPV commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,363 @@
+	}
+
+	/* Allow publishing to finish before suspending. */
+	k_sleep(K_MSEC(100));
```
Have you find the reason why it was failing without this delay?

HaavardRei replied:
I have tried debugging quite a bit, and it seems that without the delay, the advertising thread goes into a spinlock which prevents further publication. I think when suspending the [advertising work](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L65) immediately after publication (without the delay) it get preempted somewhere [here](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L287-L289). 

I have a figure comparing the log output without (left) and with (right) the delay:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/1e328274-7a60-4e16-83c3-ccba12ec769c)
 It may seem that some functionality in the access layer is disturbed by suspending immediately after a publication.

PavelVPV replied:
What if schedule the adv work right after stopping and removing adv instances if `ADV_FLAG_SCHEDULED` or `ADV_FLAG_SCHEDULE_PENDING` is set? Then it will try to send the buffer and fail, but `struct bt_mesh_send_cb.start` will be called, which will notify the access layer.

HaavardRei replied:
Fixed after offline discussion

PavelVPV commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -222,6 +222,13 @@ void bt_mesh_adv_init(void)
+
+int bt_mesh_adv_disable(void)
+{
+	k_thread_suspend(&adv_thread_data);
```
If I'm not mistaken, this will only not let schedule the adv thread, but the adv pool will not be purged, so the behavior for legacy and extended advertiser will be different.

HaavardRei replied:
That is true, is there some corresponding functionality that should be used in regards to k_threads?

PavelVPV replied:
I'm not sure, but as an idea (not the only one, of course), is to add a flag which is constantly checked in the adv thread. If the flag is set, it purges the pool like adv_ext does and then exits the thread. Then `bt_mesh_adv_enable()` starts the thread again.

HaavardRei replied:
Changed to checking a flag in the adv thread, then aborting the thread once the adv thread exits (using a semaphore to ensure the tread is not in the middle of some work). 
`bt_mesh_adv_enable()` runs the thread, but requires `bt_mesh_adv_init()` to be called beforehand to create the thread again after abortion. Changed the `bt_mesh_resume()` function to call `bt_mesh_adv_init()` if legacy advertising is used.

PavelVPV replied:
Hm, if I'm not mistaken, the adv buffers that were pushed to the adv pool but not sent yet, will stay in the pool until the thread is started again. This creates different behavior from the extended advertiser, where all buffers will be dropped and the upper layers will be notified over `bt_mesh_send_cb.start` callback immediately. Probably, we should keep the thread alive, but if the flag is set, for every adv buffer that falls into the adv pool, call `bt_mesh_adv_send_start()` with `-ENODEV` error (like for the extended advertiser if I'm not mistaken)? Between this code:
https://github.com/zephyrproject-rtos/zephyr/blob/8b58048d02015e21a8c5b71857fa4289cd1cb8da/subsys/bluetooth/mesh/adv_legacy.c#L181-L184
Something like:
```
if (!enabled) {
        bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));
        net_buf_unref(buf);
        continue;
}
```

The other option is to prevent allocating any adv buffer in `bt_mesh_adv_create` if the stack is suspended. But we need to check if return value of `bt_mesh_adv_create` is correctly handled in the stack.

HaavardRei replied:
Keeping the thread alive works when just suspending mesh, but not when disabling Bluetooth. Not joining/aborting the thread before disabling BT results in an error:
![image](https://github.com/zephyrproject-rtos/zephyr/assets/91412738/75c1687e-f6c0-4ab7-b3be-97c8970bfa46)

I do not immediately see the problem in terminating the thread, by suspending mesh we will not send anything until we resume mesh? Is there some cleanup we can instead do when the adv thread exits its while loop?

PavelVPV replied:
The problem in the different behavior of legacy and ext advertiser. As I suggested, the other option is that prevent allocating adv buffers in `bt_mesh_adv_create` if mesh is suspended. The adv buffers that are already in the pool should be released and the start callback should be called with error code.

HaavardRei replied:
I believe this is already implemented:
https://github.com/zephyrproject-rtos/zephyr/blob/934a09550dd17406426ac92aac89761cbe506520/subsys/bluetooth/mesh/adv.c#L114-L117

For releasing the adv buffers already in the pool, is it sufficient to call `bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));`, or is there something I am missing?

HaavardRei replied:
I believe this is already implemented:
https://github.com/zephyrproject-rtos/zephyr/blob/934a09550dd17406426ac92aac89761cbe506520/subsys/bluetooth/mesh/adv.c#L114-L117

For releasing the adv buffers already in the pool, is it sufficient to call `bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));`, or is there something I am missing?

HaavardRei replied:
Discussed offline: Updated to empty the adv buffers by using `bt_mesh_adv_buf_get` to empty the pool upon exiting the adv thread.

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -506,6 +506,26 @@ int bt_mesh_adv_enable(void)
 	return 0;
 }
 
+int bt_mesh_adv_disable(void)
```
Discussed offline. All adv attempts will fail here: https://github.com/zephyrproject-rtos/zephyr/blob/4fe2605160fd535980924a7ea09080c5984fe896/subsys/bluetooth/mesh/adv_ext.c#L147-L150

But do you know what happens with already scheduled advertising? Is `adv_sent` callback is called? If so, won't such advertisement be lost here: https://github.com/zephyrproject-rtos/zephyr/blob/4fe2605160fd535980924a7ea09080c5984fe896/subsys/bluetooth/mesh/adv_ext.c#L453-L458

I tend to think that we need the following test:
- Fill up the advertiser pool.
- When the first advertisement is started (`struct bt_mesh_send_cb.start` is called), call `bt_mesh_suspend()`.
- Expect that `struct bt_mesh_send_cb.end` is called for all advertisements.
- While the stack is suspended, try to schedule one more advertisement. 
- Expect that `struct bt_mesh_send_cb.start` is called with error (or what behavior do we want here?).
- Call `be_mesh_resume()` and fill up the adv pool again.
- Expect that all advertisements are sent successfully.

Feel free to create a separate task for this test as it was not planned, but I think this test will guarantee that advertiser behavior works as we want. It will also guarantee that the adv pool is emptied correctly and can be reused later. The test with publication IMHO seems very high level and don't give enough confident that suspend/resume behavior works correctly.

HaavardRei replied:
Regarding the initial comment, we arrived at the conclusion that there was no need to unref buffers. But should still `bt_mesh_adv_send_start()` be called? If so, with what error code? Is this sufficient for the "clean-up" part, or is there more that should be done?

The already scheduled advertising is cancelled, with [this line](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L514). 

I will create a task for the final part, thanks for the input!

PavelVPV replied:
> Regarding the initial comment, we arrived at the conclusion that there was no need to unref buffers. But should still `bt_mesh_adv_send_start()` be called? If so, with what error code? Is this sufficient for the "clean-up" part, or is there more that should be done?

True, it is not needed as `bt_mesh_adv_send_start()` will be called here with an error:
https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L231

> The already scheduled advertising is cancelled, with [this line](https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L514).

Sorry, by scheduled I meant pushed to the host by this call: https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L178
`adv->buf` will contain reference to that buffer (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L228), which is then unreferenced here: https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L267-L270
So the question is will it work if adv->instance is NULL (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L527)? Or is `adv_send` callback is called in `bt_le_ext_adv_stop` (https://github.com/zephyrproject-rtos/zephyr/blob/e13912d92323c6ba94665c898408a18ef4cd7f3c/subsys/bluetooth/mesh/adv_ext.c#L516) and then there is no issue?

HaavardRei replied:
Discussed offline.

HaavardRei commented:
Granted that the changes in the PR is ok, the failing "provision/pb_remote_timeout" test can be fixed by adding `k_sleep(K_MSEC(100))` before this call:
https://github.com/zephyrproject-rtos/zephyr/blob/bc08e801c6b1416dcbbc679f5e25f801d438a531/tests/bsim/bluetooth/mesh/src/test_provision.c#L159

The problem is that mesh is suspended in the same period a message is sent, interrupting the message.

PavelVPV commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -506,6 +506,33 @@ int bt_mesh_adv_enable(void)
+			return err;
+		}
+
+		adv_sent(advs[i].instance, NULL);
```
Minor: please add a comment that `adv_sent` is called to finish a transmission of an adv buffer that was pushed to the host before the advertiser was stopped, but didn't finish.

PavelVPV commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -225,6 +229,15 @@ int bt_mesh_adv_enable(void)
+int bt_mesh_adv_disable(void)
+{
+	enabled = false;
+	k_sem_take(&adv_thread_completion_sem, K_FOREVER);
```
I think you can use `k_thread_join` instead of the semaphore. Not sure if `k_thread_abort` is needed as the thread will be stopped after `k_thread_join` (unless it timed out).

PavelVPV commented:
> Granted that the changes in the PR is ok, the failing "provision/pb_remote_timeout" test can be fixed by adding `k_sleep(K_MSEC(100))` before this call:
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/bc08e801c6b1416dcbbc679f5e25f801d438a531/tests/bsim/bluetooth/mesh/src/test_provision.c#L159
> 
> The problem is that mesh is suspended in the same period a message is sent, interrupting the message.

If I understand correctly this will let the transport layer to send a SegAck. Sounds fine by me for now, though it can brake later. Probably k_work_schedule with 100ms will be better as it will not block bt rx thread and allow to process other messages (segments retransmission during this timeout).

PavelVPV commented on subsys/bluetooth/mesh/adv_legacy.c:
```diff
@@ -188,6 +189,21 @@ static void adv_thread(void *p1, void *p2, void *p3)
 	}
+
+	/* Empty the advertising pool when advertising is disabled */
+	while(true)
+	{
+		struct net_buf *buf;
+
+		buf = bt_mesh_adv_buf_get(K_NO_WAIT);
+
+		if (!buf) {
+			break;
+		}
+
+		bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));
+        net_buf_unref(buf);
+	}
```
Small improvement:
```suggestion
	while(buf = bt_mesh_adv_buf_get(K_NO_WAIT))
	{
		bt_mesh_adv_send_start(0, -ENODEV, BT_MESH_ADV(buf));
        	net_buf_unref(buf);
	}
```
And move declaration `struct net_buf *buf;` to the beginning of the function (and removing it from line 150).

HaavardRei replied:
Done!

PavelVPV commented:
You can rebase now to fix broken CI

HaavardRei committed:
Bluetooth: Mesh: advertiser: add disable function

Adds a disable function for the extended advertising which stops and
deletes the advertising instances, allowing them to be properly
reinitialized when calling `bt_mesh_adv_enable()` after bluetooth has
been disabled and re-enabled.

For the legacy advertising, the function terminates the advertising
thread. If legacy advertising is used, `bt_mesh_adv_init()` must be
called before `bt_mesh_adv_enable()` to properly resume advertising
after suspension.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
bsim: Bluetooth: Mesh: Add suspension test

Adds two test-cases; one for suspending and resuming Mesh,
and one for suspending Mesh and disabling Bluetooth before
re-enabling Bluetooth and resuming Mesh.

Both tests checks that periodic publication is stopped during
suspension.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
bsim: Bluetooth: Mesh: fix provision timeout

Adapts the `pb_remote_timeout` test to the updated `bt_mesh_suspend`
API.
Suspension is scheduled with a small delay to allow publications to be
sent before suspending Mesh.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

HaavardRei committed:
Bluetooth: Mesh: suspend/resume advertising

Disables Mesh advertising when suspending, and enables advertising
when resuming Mesh.

Signed-off-by: Håvard Reierstad <haavard.reierstad@nordicsemi.no>

PavelVPV commented:
I guess DNM can be removed?

alxelax commented on subsys/bluetooth/mesh/adv_ext.c:
```diff
@@ -506,6 +506,36 @@ int bt_mesh_adv_enable(void)
+			return err;
+		}
+
+		/* `adv_sent` is called to finished transmission of an adv buffer that was pushed to
```
```suggestion
		/* `adv_sent` is called to finish transmission of an adv buffer that was pushed to
```

alxelax commented on subsys/bluetooth/mesh/adv.h:
```diff
@@ -80,6 +80,8 @@ int bt_mesh_scan_disable(void);
 
 int bt_mesh_adv_enable(void);
 
+int bt_mesh_adv_disable(void);
```
Please add it shouldn't be called from work queue since behavior is undefined.
Probably it is worth to add to public suspension API

This is `k_work_flush_delayable` note:
```
Note
Be careful of caller and work queue thread relative priority. If this function sleeps it will not return until the work queue thread completes the tasks that allow this thread to resume.
Behavior is undefined if this function is invoked on dwork from a work queue running dwork.
```

HaavardRei replied:
Added comment to `adv.h` and a note in `bt_mesh_suspend()` API description

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,365 @@
+	prev_num = num;
+
+	/* Ensure that no message is received while Mesh is suspended or disabled. */
+	if (suspended) {
```
```suggestion
	ASSERT_FALSE_MSG(suspended, "Received publication while Mesh is suspended.");
```

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,365 @@
+	return 0;
+}
+
+static int test_msg_handler(const struct bt_mesh_model *model, struct bt_mesh_msg_ctx *ctx,
```
All functions those are not really tests shouldn't be begun with `test` prefix. Please rename.

HaavardRei replied:
Renamed!

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,365 @@
+		FAIL("AppKey add failed (err %d, status %u)", err, status);
+	}
+
+	if (rx) {
```
Just suggestion to be more compact:
```
const struct bt_mesh_test_cfg *pcfg = rx ? &rx_cfg : &tx_cfg;
uint16_t model_id = rx ? TEST_MODEL_ID_2 : TEST_MODEL_ID_1;

err = bt_mesh_cfg_cli_mod_app_bind(0, pcfg->addr, pcfg->addr, 0, model_id, &status);
if (err || status) {
	FAIL("Model %#4x bind failed (err %d, status %u)", model_id, err, status);
}
```

alxelax commented on tests/bsim/bluetooth/mesh/src/test_suspend.c:
```diff
@@ -0,0 +1,365 @@
+		}
+	}
+
+	err = bt_mesh_suspend();
```
Please use in all such places `ASSERT_OK` macro. Places where just check error code. There are many in the file. 
```
ASSERT_OK(bt_mesh_suspend());
```
It will add automatically general message with error code if it happened.
If you need additional message, just use `ASSERT_OK_MSG`
```
for (int i = 0; i < NUM_PUB; i++) {
	ASSERT_OK_MSG(k_sem_take(&publish_sem, K_SECONDS(30)), "Pub timed out");
}
```

HaavardRei replied:
Fixed, thanks for the suggestion!

alxelax reviewed (commented):
LGTM, but requires a bit more cosmetic changes.
